crates/red_knot_python_semantic/src/types/unpacker.rs

```diff
@@ -1,126 +1,185 @@
-                            Cow::Owned(element_types)
+            | ast::Expr::Tuple(ast::ExprTuple { elts, .. }) => {
+                // Initialize the vector of target types, one for each target.
+                //
+                // This is mainly useful for the union type where the target type at index `n` is
+                // going to be a union of types from every union type element at index `n`.
+                //
+                // For example, if the type is `tuple[int, int] | tuple[int, str]` and the target
+                // has two elements `(a, b)`, then
+                // * The type of `a` will be a union of `int` and `int` which are at index 0 in the
+                //   first and second tuple respectively which resolves to an `int`.
+                // * Similarly, the type of `b` will be a union of `int` and `str` which are at
+                //   index 1 in the first and second tuple respectively which will be `int | str`.
+                let mut target_types = vec![vec![]; elts.len()];
+
+                let unpack_types = match value_ty {
+                    Type::Union(union_ty) => {
+                        Either::Left(union_ty.elements(self.db()).iter().copied())
+                    }
+                    _ => Either::Right(std::iter::once(value_ty)),
+                };
+
+                for ty in unpack_types {
+                    // Deconstruct certain types to delegate the inference back to the tuple type
+                    // for correct handling of starred expressions.
+                    let ty = match ty {
+                        Type::StringLiteral(string_literal_ty) => {
+                            // We could go further and deconstruct to an array of `StringLiteral`
+                            // with each individual character, instead of just an array of
+                            // `LiteralString`, but there would be a cost and it's not clear that
+                            // it's worth it.
+                            Type::tuple(
+                                self.db(),
+                                std::iter::repeat(Type::LiteralString)
+                                    .take(string_literal_ty.python_len(self.db())),
+                            )
                         }
-                    } else {
-                        Cow::Borrowed(tuple_ty.elements(self.db()).as_ref())
+                        _ => ty,
                     };
 
-                    for (index, element) in elts.iter().enumerate() {
-                        self.unpack(
-                            element,
-                            element_types.get(index).copied().unwrap_or(Type::Unknown),
-                            scope,
-                        );
+                    if let Some(tuple_ty) = ty.into_tuple() {
+                        let tuple_ty_elements = self.tuple_ty_elements(elts, tuple_ty);
+
+                        // TODO: Add diagnostic for length mismatch
```

Comment:
Will we run into duplicate diagnostic problems if we add diagnostics here (because unpack is called multiple times) or is it guaranteed that unpack is only called once?

Reply:
I don't think we should get duplicate diagnostics because of the `first` flag:

https://github.com/astral-sh/ruff/blob/92928560f8587995003eaf2ec0730f1deb802287/crates/red_knot_python_semantic/src/types/infer.rs#L1863-L1868